Wrench support extend

nixnoxus opened the pull request:
Adds wrench support for bones, connected_chests and some nodes from pipeworks and xdecor.

For more information see: https://github.com/mt-mods/wrench/tree/wrench_support_extend#supported-mods-and-nodes

Related to https://github.com/mt-mods/technic/pull/251

nixnoxus committed:
change <world>/technic.conf:enable_wrench_crafting -> minetest.conf:wrench.enable_crafting

nixnoxus committed:
add bones support

nixnoxus committed:
add connected_chests support

nixnoxus committed:
add xdecor support

nixnoxus committed:
add pipeworks support [WIP]

nixnoxus committed:
add "technic:injector" support

nixnoxus committed:
update luacheck

nixnoxus committed:
update README.md

nixnoxus committed:
add "pipeworks:nodebreaker_*" support

nixnoxus committed:
Update README.md

crafting only in combination with technic modpack https://github.com/mt-mods/wrench/issues/1

OgelGames committed:
support node timers

OgelGames committed:
adjust craft recipe

OgelGames committed:
fixes

OgelGames committed:
`more_chests` support

OgelGames commented on nodes/xdecor.lua:
```diff
@@ -0,0 +1,112 @@
+	"xdecor:cauldron_soup",
+}
+
+local function translated(player, text)
+	local player_name = player:get_player_name()
+	local lang_code = minetest.get_player_information(player_name).lang_code
+	return minetest.get_translated_string(lang_code, text)
+end
+
+for _, nodename in pairs(nodes) do
+	wrench.register_node(nodename, {
+		metas = {
+			infotext = wrench.META_TYPE_STRING,
+		},
+		description = function(pos, meta, node, player)
+			local str = translated(player, minetest.registered_nodes[node.name].infotext)
+			local i = string.find(str, "-")
+			return (i and i > 0) and string.sub(str, 1, i - 1) or str
+		end,
+
+	})
+end
```
I don't think this is a good idea, not only will the picked up item be locked to the language of whoever picked it up, but the string manipulation might not work for some translations.

Maybe use the description of the empty cauldron?

S-S-X replied:
What's wrong with getting infotext normal way and leaving translations for xdecor itself?
Translation happens client side and normally you do not want to translate things server side unless it is strictly needed for server side processing, wrench does not do any server side processing.

nixnoxus replied:
some infotexts seem to be too long for me. see
- https://github.com/minetest-mods/xdecor/blob/master/src/cooking.lua#L191
- https://github.com/minetest-mods/xdecor/blob/master/src/cooking.lua#L213

that's why I wanted to cut off the part after '-'.
but yes, it is cleaner to solve this in xdecor by a shorter description. 
i'll adjust this accordingly ...

OgelGames commented:
I made some changes and added support for `more_chests`, I hope you don't mind, I thought it would be easier to work on the same branch so we can both work on support for different mods :)

nixnoxus committed:
remove the use of translated()

nixnoxus committed:
update README.md

nixnoxus committed:
fix node timers

nixnoxus committed:
add "pipeworks:*filter" support

OgelGames commented on functions.lua:
```diff
@@ -81,7 +81,7 @@ function wrench.pickup_node(pos, player)
 		local timer = minetest.get_node_timer(pos)
 		local timeout = timer:get_timeout()
-		if timeout > 0 then
+		if timeout then
```
I'm curious about why you changed this, `get_timeout()` always returns a value, with 0 being no timer running.

Is it different in other minetest versions? (I use 5.5.0-dev)

nixnoxus committed:
if timeout then

nixnoxus commented:
> I'm curious about why you changed this, `get_timeout()` always returns a value, with 0 being no timer running.

you are right, i fixed it
I removed the 'if timeout> = 0' so that `technic.injector` may remain `disabled` after placement

> Is it different in other minetest versions? (I use 5.5.0-dev)

i think no. i use 5.4.1

OgelGames commented:
Ah ok 👍 I guess I didn't test it enough to notice that :P

OgelGames committed:
`3d_armor_stand` support

OgelGames committed:
add missing optional deps

nixnoxus committed:
support node drop

nixnoxus committed:
drop pipeworks wielder in state "off" like digging

OgelGames commented on functions.lua:
```diff
@@ -47,6 +47,11 @@ function wrench.pickup_node(pos, player)
 			return false, errors.owned:format(owner)
 		end
 	end
+	local drop = def.drop and minetest.registered_nodes[node.name].drop
+	if drop then
+		-- drop node name like digging
+		node.name = drop
+	end
```
`drop` can also be a table: [lua_api.txt#L7861-L7914](https://github.com/minetest/minetest/blob/80c3c7e642749f9316a3eee2c235df3ce8be1666/doc/lua_api.txt#L7861-L7914).

I think it might be better to specify it in the wrench definition.
```suggestion
	if def.pickup_name then
		node.name = def.pickup_name
	end
```

nixnoxus committed:
add some "pipeworks:*tube*" support

nixnoxus commented:
> drop can also be a table: lua_api.txt#L7861-L7914.

oops. but I would like the possibility to take the drop name from the mod. how about this:

```patch
diff --git a/functions.lua b/functions.lua
index a5424e3..453fb05 100644
--- a/functions.lua
+++ b/functions.lua
@@ -47,10 +47,14 @@ function wrench.pickup_node(pos, player)
                        return false, errors.owned:format(owner)
                end
        end
-       local drop = def.drop and minetest.registered_nodes[node.name].drop
-       if drop then
+       if type(def.drop) == "string" then
+               node.name = def.drop
+       elseif type(def.drop) == "boolean" then
                -- drop node name like digging
-               node.name = drop
+               local drop = def.drop and minetest.registered_nodes[node.name].drop
+               if type(def.drop) == "string" then
+                       node.name = drop
+               end
        end
        local data = {
                name = node.name,
```
we would then have both options

OgelGames commented:
> how about this
> we would then have both options

Good idea 👍

nixnoxus committed:
`def.drop` can be a string or a boolean

nixnoxus committed:
nodes/pipeworks.lua

bugfix: call pipeworks.after_place() to connect (with) tubes correctly

nixnoxus committed:
update luacheck

OgelGames committed:
move handling for `def.drop` to `wrench.register_node()`

OgelGames committed:
add `def.after_pickup` and always update tube nodes

OgelGames committed:
move api functions to api.lua

OgelGames committed:
little code fixes

OgelGames committed:
`biofuel` refinery support

nixnoxus committed:
add setting wrench.enable_debug (very experimental)

nixnoxus committed:
add `xdecor:hive` support

nixnoxus committed:
add `bees:hive_wild` support

nixnoxus committed:
add `vessels:shelf` support

nixnoxus committed:
add `mesecons` support (WIP)

S-S-X reviewed (commented):
Add proper tests Instead of adding debug code into mod.
Optionally if debug code like that is needed locally file can be added to local .gitignore
Local debug mod is also one option for this.

nixnoxus committed:
fix mesecons disconnection

nixnoxus committed:
update luacheck

nixnoxus commented:
> Add proper tests Instead of adding debug code into mod.
> Optionally if debug code like that is needed locally file can be added to local .gitignore
> Local debug mod is also one option for this.

the debug code provides helpful information about objects that are not part of this mod (to support them). 
i played around with mineunit a bit, but haven't (yet) found a way to access objects from other mods.

we can simply remove the debug code before merge. This is why it is in debug.lua

S-S-X commented:
> played around with mineunit a bit, but haven't (yet) found a way to access objects from other mods

You either load other mod (not recommended) or add fixtures (recommended)

nixnoxus committed:
add `digilines` support

nixnoxus committed:
add `mobs:spawner` support

OgelGames committed:
also update connections on place

OgelGames committed:
meta:get_string always returns a string

OgelGames committed:
fix metas

S-S-X commented:
> > played around with mineunit a bit, but haven't (yet) found a way to access objects from other mods
> 
> You either load other mod (not recommended) or add fixtures (recommended)

I guess I could add simple Mineunit API tests which should give some idea about fixtures.
Could as well try to improve Mineunit tool use a bit... was one of things I've been working on anyway, while usable it is bit too verbose and not exactly intuitive currently.

nixnoxus committed:
use `drop_node` in get_description()

nixnoxus committed:
fix luacontroller* and descriptions added

nixnoxus committed:
add `realchess:chessboard` support

nixnoxus committed:
fix typo

OgelGames commented on nodes/mesecons.lua:
```diff
@@ -69,10 +82,14 @@ for d = 0, 1 do
 end
 end
 
+luacontroller_defs.drop = nil
```
Wouldn't it be better to change it to an unburnt Luacontroller? 🤔 (like the Lua tube)

S-S-X replied:
If wrench is meant to allow moving nodes unmodified then I think no but it is more about making decision what wrench is.

S-S-X replied:
Moving as is burnt is burnt allows better compatibility with some corner cases and afaik there's no losses: memory is already cleared and luac can be jus reprogrammed.
Really corner cases like customization mod that requires some items for repairing burnt lua controller, bit far fetched but as decision how tool should consistently behave it might be better.

nixnoxus commented:
sorry, for resolving. I clicked wrong:(

I also find it more logical to give the burned ones. that's what the wrench is for.

nixnoxus commented:
`pipeworks: broken_tube_1` should maybe too can be taken with the wrench. what do you all mean?

OgelGames commented:
I think broken tubes should not be able to be picked up, as they are a placeholder for multiple different tubes.

nixnoxus committed:
rename functions "wrench.*" to "wrench:*" (#3)

nixnoxus committed:
fix crash when `wrench.enable_debug = true`

nixnoxus committed:
fix typo in description for `realchess:chessboard`

OgelGames committed:
Revert "rename functions "wrench.*" to "wrench:*" (#3)"

OgelGames committed:
add compatibility for `:`

fixes #3

OgelGames committed:
Revert "add compatibility for `:`"

This reverts commit 302173124ade4d433c1a7f443c158e03e14e0e7f.

OgelGames committed:
add `wrench.plus = true`

https://github.com/mt-mods/wrench/issues/3#issuecomment-986166887

OgelGames commented:
I realized there's two problems with picking up burnt Luacontrollers; they don't have a description, and if `wrench` was later disabled on a server, players would be left with an "illegal" node in their inventory.

~~If there would be one rule to how the wrench should behave, it would be that it shouldn't put unobtainable items in a player's inventory.~~

The lack of description is the more obvious issue though:
![image](https://user-images.githubusercontent.com/48543043/144735616-e14ee1d3-b4a6-4801-b619-95e54a671f23.png)

EDIT: actually that would mean the colored technic chests would be not allowed, so that isn't a possible rule...

OgelGames commented:
> I realized there's two problems with picking up burnt Luacontrollers; they don't have a description, and if `wrench` was later disabled on a server, players would be left with an "illegal" node in their inventory.
> 
> If there would be one rule to how the wrench should behave, it would be that it shouldn't put unobtainable items in a player's inventory.
> 
> The lack of description is the more obvious issue though: ![image](https://user-images.githubusercontent.com/48543043/144735616-e14ee1d3-b4a6-4801-b619-95e54a671f23.png)

nixnoxus commented:
> If there would be one rule to how the wrench should behave, it would be that it shouldn't put unobtainable items in a player's inventory.

Maybe control this via `privs`? Then the server admin has control over it. Like:

```diff
diff --git a/nodes/mesecons.lua b/nodes/mesecons.lua
index e3f406e..23a39b8 100644
--- a/nodes/mesecons.lua
+++ b/nodes/mesecons.lua
@@ -92,4 +92,5 @@ end
 end
 
 luacontroller_defs.drop = nil
+luacontroller_defs.privs = "wrench_special_nodes"
 wrench.register_node("mesecons_luacontroller:luacontroller_burnt", luacontroller_defs)
```

Alternative `wrench.special_nodes = true` as setting.

This also affects `bees:hive_wild`.

OgelGames committed:
fixes for lua nodes

OgelGames commented:
I realized that's not a possible rule because of colored technic chests, those have always been able to be picked up, so that cannot change.

I fixed the Luacontroller and Lua tube to take their description from their unburnt nodes.

nixnoxus commented:
@OgelGames: do you have a special mod for testing or do you also test manually?

OgelGames commented:
I test manually. Even though some things are missed occasionally, I still find it better than any automated tests.

nixnoxus committed:
Update README.md

nixnoxus committed:
list supported modnames and nodes

nixnoxus committed:
Update README.md

nixnoxus commented:
the translations do not work for me with this code:
```lua
local S = rawget(_G, "intllib") and intllib.Getter() or function(s) return s end
```
I already noticed that with the technic mod. Does anyone against it if I rewrite it? Like: 
```lua
local S = minetest.get_translator()
```

OgelGames commented:
Go ahead, it was already planned to be done: https://github.com/mt-mods/technic/issues/104 👍

nixnoxus committed:
fix and update translations

nixnoxus committed:
fix luacheck warning

OgelGames committed:
modname is always "wrench"

nixnoxus commented:
FYI: I converted and updated the locale/ directory with https://github.com/nixnoxus/minetest_hacks/blob/main/bin/update-locale.sh 

There is already a script for updating locale: https://github.com/minetest-tools/update_translations
But it removes comments at the beginning (from the translators) and destroys the structure.

nixnoxus committed:
Revert "Revert "add compatibility for `:`""

This reverts commit 820b8d5b60432fa0555f582dd9a5365ae9c8a833.

nixnoxus commented:
i think it is time for review and merge. :rocket:

OgelGames committed:
pass `pointed_thing` to `def.after_place`

OgelGames committed:
`signs_lib` support

OgelGames committed:
only override nodes once

OgelGames committed:
mark log messages with "[wrench]"

OgelGames committed:
luacheck and depends

OgelGames commented:
> i think it is time for review and merge. 🚀

Not quite yet, still got a few more mods to add support for 😉 

Also, it might be good to split the `mesecons_*` nodes into each mod, as they *could* be disabled (because `mesecons` is a modpack).

OgelGames committed:
`basic_signs` support

OgelGames committed:
fix glow items being dropped from signs

nixnoxus committed:
add wrench.description_with_configuration()

nixnoxus committed:
split `mesecons_*`

nixnoxus committed:
fix `mesecons_microcontroller:microcontroller*` drop

nixnoxus committed:
update german translation

nixnoxus committed:
use wrench.description_with_configuration() when `lists` and `description` are not specified

nixnoxus committed:
add `protector` support

nixnoxus committed:
restore nodes/pipeworks.lua

nixnoxus committed:
fix spaces

S-S-X commented on api.lua:
```diff
@@ -0,0 +1,53 @@
+	end
+end
+
+minetest.register_on_mods_loaded(function()
+	for name, def in pairs(wrench.registered_nodes) do
+		local node_def = minetest.registered_nodes[name]
+		if node_def then
+			local old_after_place = node_def.after_place_node
+			minetest.override_item(name, {
+				after_place_node = function(...)
+					if not wrench.restore_node(...) and old_after_place then
+						return old_after_place(...)
+					end
+				end
+			})
+		else
+			minetest.log("warning", "[wrench] Registered node is now unknown: "..name)
+			wrench.registered_nodes[name] = nil
+		end
+	end
+end)
```
I'd recommend reverting this change to registration API unless there's very good reason to execute it only after every other mod is loaded (remember that all dependencies are loaded already).

It does:
* prevent registering nodes runtime using api
* makes override ordering a lot harder to predict

OgelGames replied:
It does three things:

- Ensures the override only happens once, even if a node is registered multiple times.
- Ensures that it's the last override, regardless of what mods override the callback.
- Allows unregistering wrench support for a node, before the override happens.

I don't know of any other way to do that.

S-S-X replied:
> Ensures the override only happens once, even if a node is registered multiple times.

Doing override multiple times does not matter, it is same as unregistering and registering again. Could just override previous override, should not be anything wrong with that.

> Ensures that it's the last override, regardless of what mods override the callback.

No, it does not. It just makes it more likely that it is last override done from said callback for mods that do not depend on wrench. On the other hands for mods that do depend on wrench... well you know already.

> Allows unregistering wrench support for a node, before the override happens.

This is nice thing (is there actual use cases?). However having it available runtime can allow unregistering wrench support any time. Maybe think for a while if there would be a better way to provide this functionality.

Real override unregister/rollback can be done, that should not be problem but this being one of reasons why registration is delayed and made all at once begs questions: how and why is unregistering used currently?

> I don't know of any other way to do that.

Not sure if I know how to do that but it is very hard to imagine that delayed registration is really needed, provided points at least are not really supporting it but also I've not worked on this so do not know if there actually is something that makes delayed registration work better.

OgelGames replied:
I tried some other ways of implementing this (caching the original callbacks, checking if the node had been overridden), but I couldn't find a way to cleanly implement everything; delaying the registration is the best solution I know of.

As an example, imagine some mods A, B, C and D, with immediate registration: 
- A registers a node.
- B registers wrench support for the node from mod A. (override happens, wrapping `after_place_node`)
- C, unaware of mod B, also adds wrench support for the node. (`after_place_node` is now double wrapped)
- D, unaware of B and C, overrides the node from mod A. (`after_place_node` gets replaced, so there's no wrapping for wrench functionality)

If the override (wrapping of `after_place_node`) gets delayed until all mods have loaded, then neither of the two problems happen.

S-S-X replied:
I think there's no reason to account for _"B registers wrench support for the node from mod A"_, in that case A and B should be well aware of each other if it is modpack, server customization mod or game. If it is just basic mod then B should not be doing that (if it handles stuff in A it should already know what A already does).
Same for _"C, unaware of mod B, also adds wrench support for the node."_, it simply should not be doing that ever without knowing about node it is registering third party support (it is server specific customization, modpack or game if it does that with valid reason).

For **any** double wrapping it should be possible to simply check if node override is already done or not.

But for _"D, unaware of B and C, overrides the node from mod A."_:
This probably could be simpler put as _"D overrides after_place_node from mod A which registered node for wrench"_ which is scenario where D should know about wrench but probably wont take wrench internal mechanisms into account (and should not have to worry about it).
That causes situation where override has to be delayed (without having to consider B or C at all)?
For this there probably is no good solutions around it without delaying registration.

S-S-X replied:
Well, only solution for that would be to drop API from wrench and only registering support internally but that would mean no API and therefore nobody else being able to add wrench support for nodes, that does not sound like good solution (it has upsides but would deny too much in my opinion).

OgelGames replied:
"*A registers a node.*" means `minetest.register_node()`, not registering for wrench ;)

S-S-X replied:
> "_A registers a node._" means `minetest.register_node()`, not registering for wrench ;)

Well, then you only have to consider B and D and instead of A and D. Other than that scenario seems to stay unaffected.

nixnoxus committed:
initial api.txt

S-S-X commented on api.txt:
```diff
@@ -0,0 +1,56 @@
+        -- Default
+        after_place = function(pos, player, itemstack, pointed_thing),
+        -- Function called after place node
+    }
```
Maybe make it md instead of plain txt from beginning? For better formatting: API.md

nixnoxus committed:
api.txt -> API.md

nixnoxus committed:
fix basic_signs: +metas.unifont

nixnoxus committed:
update README.md

nixnoxus committed:
fix signs_lib: +metas.unifont

nixnoxus committed:
Update API.md

nixnoxus committed:
Update API.md

OgelGames committed:
add `wrench.unregister_node` and move drop check

OgelGames committed:
add `wrench.description_with_text`

nixnoxus committed:
change `has_*` to `wrench.has_*`

nixnoxus committed:
*fix `technic:cnc_mk2` and `technic:*_battery_box*`: +metas.channel

nixnoxus committed:
add `technic:forcefield_emitter_*` and `technic:quarry`

nixnoxus committed:
fix strip_esc()

nixnoxus committed:
add `wrench.description_with_channel`

nixnoxus committed:
check `meta` and `lists` before pickup when `wrench.enable_debug = true`

nixnoxus committed:
add `digistuff` support (WIP)

nixnoxus committed:
add `wrench.META_TYPE_IGNORE`

nixnoxus committed:
more `digistuff` support (WIP)
+ controller_programmed, eeprom, ram, ioexpander_*, piezo, timer, wall_knob_configured

nixnoxus committed:
fix: meta:to_table() before minetest.remove_node(pos)

nixnoxus committed:
even more `digistuff` support (WIP)
+ movestone, noteblock, piston*

nixnoxus committed:
update README.md

OgelGames commented on functions.lua:
```diff
@@ -147,9 +147,10 @@ function wrench.pickup_node(pos, player)
 	minetest.remove_node(pos)
 	if def.after_pickup then
-		def.after_pickup(pos, node, meta:to_table(), player)
+		def.after_pickup(pos, node, meta_table, player)
```
Why an extra variable? How is that any different?

OgelGames committed:
fix digilines lcd and remove unnecessary metas

OgelGames commented on init.lua:
```diff
@@ -6,6 +6,7 @@ wrench = {
 	plus = true,
 	registered_nodes = {},
 	blacklisted_items = {},
+	META_TYPE_IGNORE = 0,
```
What is the purpose of this? The only places it's used doesn't seem necessary. (it can be replaced by `nil`)

nixnoxus commented:
> Why an extra variable? How is that any different?

`meta:to_table()` is empty after `minetest.remove_node(pos)`
That's why I save the table before in meta_table.

> What is the purpose of this? The only places it's used doesn't seem necessary. (it can be replaced by nil)

This is the reason: https://github.com/mt-mods/wrench/pull/2/commits/3b8bf58b45e0a2be418f3a43de839042a8699459
For testing new nodes, I had installed this check. It prevents (if `wrench.enable_debug = true`) the pickup nodes with unknown `metas` or `lists`.
This is to prevent `metas` or `lists` from being missed when adding new nodes.
If the value is `nil`, the key no longer exists and cannot be checked.

OgelGames commented:
> `meta:to_table()` is empty after `minetest.remove_node(pos)`
> That's why I save the table before in meta_table.

Ah, of course 🤦

nixnoxus committed:
complete `digistuff` support
+ *touchscreen, gpu, light_*, nic

nixnoxus committed:
add `digiscreen` support

nixnoxus committed:
fix `bones:bones` with `mapserver_mod` (see: https://github.com/minetest-mapserver/mapserver_mod/blob/3.1/bones.lua#L15)

nixnoxus committed:
add misc `wrench.META_TYPE_IGNORE` keys

nixnoxus committed:
add meta `internal_EU_charge_max` for technic:*_battery_box*

nixnoxus committed:
add meta `tube_time` for technic:cnc_mk2

nixnoxus committed:
add `technic:power_monitor` and `technic:switching_station` when `wrench.has_digilines`

nixnoxus committed:
add `technic:supply_converter`

nixnoxus committed:
Update README.md

nixnoxus committed:
add `digiline_detector_tube_*`

nixnoxus committed:
add `technic:*_generator*`

nixnoxus committed:
ignore `tube_time` and `formspec` on LV base machines

nixnoxus committed:
fix `technic:*_battery_box_*`: +metas.tube_time

nixnoxus committed:
add `def.lists_ignore` (used when `wrench.enable_debug = true`)

nixnoxus committed:
fix `metas` and add `lists_ignore` for `technic:*_chest`

BuckarooBanzay commented:
the commits are piling up on this one, is there a plan to merge this someday? (i'm just asking, no pressure)

OgelGames commented:
Yeah... It really should be merged... :P

I think there are too many mods we could support to keep this open, so I think we can leave those for other PRs. I'll try to find some time to go over the "recent" additions and put a review in.

One question: Are we planning on keeping the debug stuff or removing it before merging? I don't really mind either way, keeping it just means more code to maintain. (also it's usefulness is limited IMO)

nixnoxus commented:
I would like to keep the debug code, this helped me enormously with adding and testing new and old nodes.

Maybe we can also merge https://github.com/mt-mods/wrench/pull/4 into it. I would like to add tests for the API. But I could also do that in a separate PR.

In the next few days I will create a nice screenshot with all supported nodes.

I have one question about the authors in https://github.com/mt-mods/wrench/blob/wrench_support_extend/README.md#contributors-originally-technic-modpack
Not all authors of technic have contributed to the wrench:
```
~/minetest.git/mods/.technic.mt-mods$ git remote -v
origin	https://github.com/mt-mods/technic (fetch)
origin	https://github.com/mt-mods/technic (push)
~/minetest.git/mods/.technic.mt-mods$ git log --pretty=format:"%an" | sort -u > a
~/minetest.git/mods/.technic.mt-mods$ git log --pretty=format:"%an" $(find wrench/) | sort -u > b
~/minetest.git/mods/.technic.mt-mods$ diff -u a b | grep -i -e kpoppel -e Nekogloop -e Ekdohibs -e  Nore -e ShadowNinja -e VanessaE -e BuckarooBanzay -e OgelGames -e int-ua -e groxxda -e SwissalpS
 BuckarooBanzay
-Ekdohibs
-kpoppel
-Luke aka SwissalpS
 OgelGames
 ShadowNinja
-VanessaE
```
Do we remove them (starting with '-') from the contributors  list?

S-S-X commented:
> I would like to keep the debug code, this helped me enormously with adding and testing new and old nodes.

In my opinion debug code like that should not be included with mod but instead included through some other code snippet collection / debug repository / debug mod / just locally with git ignores added. Not 100% sure but it seems like that file could just be separate mod.

If it stays then at least make sure that there is separate commit for anything related to debug thing.

> Maybe we can also merge #4 into it. I would like to add tests for the API. But I could also do that in a separate PR.

Not sure what you meant with this but I would recommend handling at least merges through separate PR possibly continuing on top of #4 or #5 instead of merging tests into this branch.
Mostly because it is easier to handle cleaning up commits to get history that allows clean cherry picks, exclusion, filtering, merging, reverting and what not.

> In the next few days I will create a nice screenshot with all supported nodes.

I would again recommend doing another branch and pull request for that instead of including those here, again makes reviews and cleaning up history a lot simpler.

> I have one question about the authors

I'd say again would be better to make another PR for that and discuss it there.
Also while commits do not always equal to contributors it might be true here because stuff is coming from technic mod, but that's just another reason to have separate PR for that.

...unless file was added here, then it makes sense to clean up before adding nonexistent contributors to master branch but just piling up unrelated stuff to PR commit list does not make sense and is not exactly helpful in long term development.

nixnoxus commented:
>> I would like to keep the debug code, this helped me enormously with adding and testing new and old nodes.

> In my opinion debug code like that should not be included with mod but instead included through some other code snippet collection / debug repository / debug mod / just locally with git ignores added. Not 100% sure but it seems like that file could just be separate mod.

In my eyes, besides tests, debug code also belongs in the respective project.
Debug.lua could be swapped out, but the sense of `wrench.META_TYPE_IGNORE` and `lists_ignore` in `nodes/*.lua` would no longer be obvious. Both variables prevent (if `wrench.enable_debug = true`) pick up nodes with unknown `meta.fields` or `inventory`s. With this I had found among others these missing `metas`:
- https://github.com/mt-mods/wrench/pull/2/commits/1c7d0895a5613bc36bd8d474d520a38db8a9fad8#diff-adfdaef00d75544242b6d93abb39ef7e74f1f379a6878ea01feff1ff202038daR8
- https://github.com/mt-mods/wrench/pull/2/commits/989fccb9d8b3126ae87a1e0d4963ba95c1e7114f#diff-783d1f08873b6e6282a9c05b0e77438b90a74f96b66200b64f48daf19179a2d6R35
- https://github.com/mt-mods/wrench/pull/2/commits/03c6bfcf5ffdd383c228de2d6d31e8ded3e17f52#diff-ea7c22a9ce0a99c97f7f792921e79ee827da6a9966d44c6cde59fd19b2b492c0R33

On the other hand, we can also run this check in general and move it to `wrench.pickup_node`. This would prevent the possible loss of meta data after updating other mods. In this case I would do without debug.lua.
 

> If it stays then at least make sure that there is separate commit for anything related to debug thing.

>> Maybe we can also merge #4 into it. I would like to add tests for the API. But I could also do that in a separate PR.

> Not sure what you meant with this but I would recommend handling at least merges through separate PR possibly continuing on top of #4 or #5 instead of merging tests into this branch.
> Mostly because it is easier to handle cleaning up commits to get history that allows clean cherry picks, exclusion, filtering, merging, reverting and what not.

Okay.

>> In the next few days I will create a nice screenshot with all supported nodes.

> I would again recommend doing another branch and pull request for that instead of including those here, again makes reviews and cleaning up history a lot simpler.

Okay.

>> I have one question about the authors

> I'd say again would be better to make another PR for that and discuss it there.
Also while commits do not always equal to contributors it might be true here because stuff is coming from technic mod, but that's just another reason to have separate PR for that.

> ...unless file was added here, then it makes sense to clean up before adding nonexistent contributors to master branch but just piling up unrelated stuff to PR commit list does not make sense and is not exactly helpful in long term development.

This is true. In this commit https://github.com/mt-mods/wrench/pull/2/commits/803d39166a233740d32dd565807020aa4e5db6b9 I had copied and pasted the authors and license from technic.
That's why I would clean it up in this PR or generally leave it like this.

S-S-X commented:
> In my eyes, besides tests, debug code also belongs in the respective project.

Can it be moved to subdir `debug` for example and move all relevant stuff from main mod in a way that everything keeps working if that sub directory is removed and all advanced debug related stuff is contained within subdirectory?
That's exactly how tests are working, there's no test related code in main mod at all, not single line or word and for some mods spec directory is simply stripped out from release package using git export rules.

nixnoxus committed:
move sanity checks to `wrench.pickup_node`

nixnoxus committed:
add misc `wrench.META_TYPE_IGNORE` keys

nixnoxus committed:
move debug.lua to debug/debug.lua

nixnoxus committed:
Update API.md

nixnoxus committed:
rename `before_remove` to `before_pickup`

nixnoxus committed:
Update README.md (cleanup #Contributors)

nixnoxus committed:
make `debug/debug.lua` optional

OgelGames commented:
I had a thought: is it possible to have mineunit perform the same checks as the debug code does? That would not only make the tests more useful, but also `debug.lua` would no longer be needed.

OgelGames committed:
Update digistuff.lua

S-S-X commented:
> I had a thought: is it possible to have mineunit perform the same checks as the debug code does? That would not only make the tests more useful, but also `debug.lua` would no longer be needed.

If Minetest engine is able to do it then Mineunit should be able to do it. However not everything is implemented for Mineunit yet as I've added functionality when it is needed.
Should take closer look at debug.lua but if it is some simple metadata checking then should be able to do that.

Biggest problem for fully automated integration testing is probably mod dependencies, Mineunit is able to load multiple mods _(as you can see from example #5 which downloads drawers as dependency)_ but in my opinion that's not really good way to do regression testing.

Mostly because it is not that easy to decide which version of other mod should be used and if latest dev/git master then it would be testing and reporting possible errors in other mods.
But also because mods needs to be downloaded for every test run, it should be possible but depending on what that requires from other mods it might not be good idea.

Fixtures can be used for proper API behavior testing against certain metadata and definitions, that's preferred way to do self contained regression testing. With this approach you basically create minimal "perfect" version of supported node by using `minetest.register_node` something like this:
https://github.com/mt-mods/wrench/blob/b105f31d59bbc4176afb1bbbf22316ef2711041f/spec/fixtures/bitchange.lua#L6-L25

OgelGames commented:
> it is not that easy to decide which version of other mod should be used

Well AFAIK we are only supporting the versions listed in the readme, so wouldn't those versions be used for testing too?

https://github.com/mt-mods/wrench/tree/wrench_support_extend#supported-mods-and-nodes

> Fixtures can be used for proper API behavior testing against certain metadata and definitions, that's preferred way to do self contained regression testing. With this approach you basically create minimal "perfect" version of supported node

The problem with that is if something changes in a mod (adding or changing metadata names), we won't know about it until someone reports it. :/

nixnoxus commented:
The debug/debug.lua does not contain classic test code but:
- Lists after loading the supported nodes
- shows code to add new nodes when wrench is used with 'sneak'.
- may return a message if a node can't be pickup (not supported or node has no meta data)

S-S-X commented:
> Well AFAIK we are only supporting the versions listed in the readme, so wouldn't those versions be used for testing too?

> The problem with that is if something changes in a mod (adding or changing metadata names), we won't know about it until someone reports it. :/

Your 2 statements contradicts with each other...

nixnoxus commented:
I think that unit tests can only ensure a stable and consistent API.

The interaction with other mods we can only ensure with integration tests (with loading other mods).

S-S-X commented:
Could possibly try separate integration tests with actual player interaction by actually loading mods and writing simple test loop where player picks up and places whitelisted nodes.
Similar to node placement test for Technic mod https://github.com/mt-mods/technic/blob/7e466e9b41215306b220d83b932b754294f32e1d/technic/spec/nodes_spec.lua#L44-L46
And placement_test function: https://github.com/mt-mods/technic/blob/7e466e9b41215306b220d83b932b754294f32e1d/technic/spec/nodes_spec.lua#L15-L20

Probably need to implement some missing functionality for Mineunit and allow disabling test failures that happen because of deprecated Minetest Lua API use, many mods are using deprecated functions and currently Mineunit throws errors for some if deprecated version is used.

nixnoxus commented:
@OgelGames i found some issues with your changes:

digistuff:detector 
https://github.com/mt-mods/wrench/pull/2/commits/29a48575363ba7189e8c44be937f71095f7aa57b#diff-0de17c9b8ddfd83bc350c98c87ecc8490af2647c2c963e15d08ec519fac173c9L84-R77
=> config loss after replace

digistuff:eeprom
https://github.com/mt-mods/wrench/pull/2/commits/29a48575363ba7189e8c44be937f71095f7aa57b#diff-0de17c9b8ddfd83bc350c98c87ecc8490af2647c2c963e15d08ec519fac173c9L107-R102
=> right click dosn't work after replace

~~digistuff:piezo
https://github.com/mt-mods/wrench/pull/2/commits/29a48575363ba7189e8c44be937f71095f7aa57b#diff-0de17c9b8ddfd83bc350c98c87ecc8490af2647c2c963e15d08ec519fac173c9L216
=> can't pickup after `digiline_send("beep", "longbeep")`~~

OgelGames commented:
> digistuff:detector => config loss after replace

It works fine for me...

> digistuff:eeprom => right click dosn't work after replace

Ah... that doesn't set it's formspec the same was as `digistuff:ram` does... 

Fixed here: https://github.com/mt-mods/digistuff/commit/b60d9b3f09555c7064ef0b3dc690e842588ad8dc

nixnoxus commented:
>>  digistuff:detector => config loss after replace

> It works fine for me...

~~My test setup:~~
~~1. place `digistuff:detector`~~
~~2. set `channel` = `c_ob`, `radius` = `1`~~
~~3. pickup `digistuff:detector`~~
~~4. (re)place `digistuff:detector`~~
~~5. right click `digistuff:detector` ~~
~~ => `channel` is now `0`~~
~~same effect also with https://github.com/mt-mods/digistuff/commit/b60d9b3f09555c7064ef0b3dc690e842588ad8dc~~

Sorry, my mistake

nixnoxus committed:
fix `digistuff:detector` channel = wrench.META_TYPE_STRING

nixnoxus committed:
fix `digistuff:movestone` state & active = wrench.META_TYPE_IGNORE

nixnoxus commented:
Looks good to me so far.

I think it would be good if we include support for `digistuff:advtouchscreen` again since it is still included in the upstream digistuff mod.  :thinking: 
(support remove here https://github.com/mt-mods/wrench/pull/2/commits/29a48575363ba7189e8c44be937f71095f7aa57b#diff-0de17c9b8ddfd83bc350c98c87ecc8490af2647c2c963e15d08ec519fac173c9L276-L282)

OgelGames committed:
Update digiscreen.lua

OgelGames committed:
whitespace and typo fixes in API.md

OgelGames committed:
copy formspec of digistuff detector and camera

because of changes in https://github.com/mt-mods/digistuff/commit/a49e4dd

OgelGames committed:
Update mesecons_detector.lua

OgelGames commented:
> I think it would be good if we include support for digistuff:advtouchscreen again since it is still included in the upstream digistuff mod.

IMO we should only support the versions stated in the readme, the API can be used for adding support for other versions/nodes. While it would be good to support more, I think in the long run it will be a lot easier to manage if we only support specific versions.

OgelGames reviewed (approved):
There is still stuff I want to add, but I don't have time for that at the moment, and it's better to make another PR anyway.

I think everything is good enough to merge 👍

nixnoxus commented:
@S-S-X please review.

S-S-X commented:
Picking up bones: `functions.lua:150: bad argument #1 to 'pairs' (table expected, got nil)`
Because placed bones do not have inventory, same situation can happen with any other node placed by other mods or direct commands, should be checked.

So no for blacklisting or whitelisting certain names but checking if that specific node has inventory and prevent pickup if it does not have, I would also recommend against digging said node because that can increase complexity and not exactly what wrench should be doing.

Simply do not do anything, maybe chat message telling that node cannot be picked up with wrench because it does not have inventory.

nixnoxus committed:
fix crash when node has no inventory

nixnoxus commented:
> Picking up bones: functions.lua:150: bad argument #1 to 'pairs' (table expected, got nil)

Fixed.

> Simply do not do anything, maybe chat message telling that node cannot be picked up with wrench because it does not have inventory.

A chat message 'Cannot pickup node. Node has no inventory.' is written.

nixnoxus commented:
if no objections come, I will merge this weekend and create a new branch / merge request for further customization.

S-S-X reviewed (approved):
Seems "mostly" fine... mostly because I've actually only read at most half of it, you know it is huge PR with over 2200 LoC added...

OgelGames commented:
It's been two months since I approved this, I think we should just merge this now.

OgelGames commented:
I don't know why github deleted the branch automatically, but I'm gonna leave it to exist for a little bit... :)
